File: scripts/diagnostics/plot_linrood_la_footprint.jl
#!/usr/bin/env julia
# ---------------------------------------------------------------------------
# Plot the 6-hour LinRood LA footprint saved by
# `linrood_la_footprint.jl`. Renders a global lat/lon map (CS panels
# unstitched onto a single equirectangular plot) showing log|dJ/dE|
# with the LA receptor marked.
# ---------------------------------------------------------------------------

using CairoMakie
using Printf

include(joinpath(@__DIR__, "..", "..", "src", "AtmosTransport.jl"))
const AT = AtmosTransport

const FOOTPRINT_PATH = joinpath(@__DIR__, "..", "..",
                                "artifacts", "linrood_la_footprint_c24_6h.bin")

function _load_footprint(path)
    open(path, "r") do io
        Nc      = read(io, Int64)
        Nz      = read(io, Int64)
        nsteps  = read(io, Int64)
        la_pan  = read(io, Int64)
        la_i    = read(io, Int64)
        la_j    = read(io, Int64)
        panels = ntuple(_ -> read!(io, Array{Float64, 2}(undef, Nc, Nc)), 6)
        return (Nc=Nc, Nz=Nz, nsteps=nsteps,
                la_panel=la_pan, la_i=la_i, la_j=la_j,
                panels=panels)
    end
end

function main()
    isfile(FOOTPRINT_PATH) || error("Footprint binary missing — run linrood_la_footprint.jl first")
    fp = _load_footprint(FOOTPRINT_PATH)
    @printf("Loaded C%d footprint (nsteps=%d) — LA at panel %d, (%d, %d)\n",
            fp.Nc, fp.nsteps, fp.la_panel, fp.la_i, fp.la_j)

    mesh = AT.CubedSphereMesh(Nc=fp.Nc, Hp=3, FT=Float64)

    # Gather per-cell (lon, lat, dJ/dE) for all 6 panels.
    lons_all = Float64[]
    lats_all = Float64[]
    vals_all = Float64[]
    for p in 1:6
        lons, lats = AT.Grids.panel_cell_center_lonlat(mesh, p)
        for j in 1:fp.Nc, i in 1:fp.Nc
            push!(lons_all, lons[i, j])
            push!(lats_all, lats[i, j])
            push!(vals_all, fp.panels[p][i, j])
        end
    end

    # Wrap longitudes to [-180, 180] for a more readable map.
    lons_wrapped = [l > 180 ? l - 360 : l for l in lons_all]

    abs_vals = abs.(vals_all)
    vmax = maximum(abs_vals)
    log_vals = [v > 0 ? log10(v / vmax) : -Inf for v in abs_vals]
    # Mask very weak cells for the map; keep the top ~3 decades.
    mask = log_vals .> -3
    @printf("Non-trivial cells (|dJ/dE| > vmax/1e3): %d / %d\n",
            count(mask), length(mask))
    @printf("Peak |dJ/dE| = %.4e\n", vmax)

    # Get the LA receptor lat/lon.
    la_lons, la_lats = AT.Grids.panel_cell_center_lonlat(mesh, fp.la_panel)
    la_lon = la_lons[fp.la_i, fp.la_j]
    la_lat = la_lats[fp.la_i, fp.la_j]
    la_lon > 180 && (la_lon -= 360)

    fig = Figure(size=(1400, 800))
    ax = Axis(fig[1, 1],
              title = @sprintf("6-hour LinRood adjoint footprint, LA receptor at (%.2f°, %.2f°)",
                                la_lat, la_lon),
              xlabel = "Longitude [°E]",
              ylabel = "Latitude [°N]",
              limits = (-180, 180, -90, 90))

    # Coastline-free background: just a continent box.
    sc = scatter!(ax, lons_wrapped[mask], lats_wrapped_indices(lats_all, mask);
                  color = log_vals[mask],
                  colormap = :viridis,
                  colorrange = (-3, 0),
                  markersize = 14)
    Colorbar(fig[1, 2], sc, label = "log₁₀(|dJ/dE| / max)")

    # Mark the LA receptor.
    scatter!(ax, [la_lon], [la_lat]; color = :red, markersize = 22,
             marker = :xcross, strokewidth = 2, strokecolor = :black)
    text!(ax, la_lon + 2, la_lat + 2; text = "LA receptor",
          color = :red, fontsize = 14)

    out_path = joinpath(@__DIR__, "..", "..", "artifacts",
                        "linrood_la_footprint_c24_6h.png")
    save(out_path, fig; px_per_unit = 2)
    @printf("Plot saved to %s (%.1f KB)\n", out_path,
            filesize(out_path) / 1024)
end

# Mirror the latitudes via the same mask for the scatter call.
lats_wrapped_indices(lats, mask) = lats[mask]

main()

File: scripts/diagnostics/plot_linrood_la_footprint.py
#!/usr/bin/env python3
"""Plot the 6-hour LinRood LA footprint from the Julia-exported CSV
(lon/lat computed by AtmosTransport so they match the production CS
geometry exactly)."""

import csv
import sys
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np

CSV_PATH = Path(__file__).resolve().parents[2] / "artifacts" / "linrood_la_footprint_c24_6h.csv"


def main():
    if not CSV_PATH.exists():
        sys.exit(f"CSV missing: {CSV_PATH}\n"
                 f"Run scripts/diagnostics/export_linrood_la_footprint_csv.jl first.")

    meta = {}
    rows = []
    with CSV_PATH.open() as f:
        header = None
        for line in f:
            line = line.strip()
            if line.startswith("#"):
                for kv in line.lstrip("# ").split():
                    if "=" in kv:
                        k, v = kv.split("=", 1)
                        meta[k] = int(v)
                continue
            if header is None:
                header = line.split(",")
                continue
            parts = line.split(",")
            rows.append((
                int(parts[0]), int(parts[1]), int(parts[2]),
                float(parts[3]), float(parts[4]), float(parts[5]),
            ))

    arr = np.array(rows, dtype=[
        ("panel", "i8"), ("i", "i8"), ("j", "i8"),
        ("lon", "f8"), ("lat", "f8"), ("v", "f8"),
    ])

    Nc = meta.get("Nc", 24)
    nsteps = meta.get("nsteps", 48)
    la_panel = meta.get("la_panel", 4)
    la_i = meta.get("la_i", 5)
    la_j = meta.get("la_j", 22)

    print(f"Loaded {len(arr)} cells (C{Nc}, nsteps={nsteps}, LA at panel {la_panel} ({la_i},{la_j}))")

    abs_v = np.abs(arr["v"])
    vmax = float(abs_v.max())
    print(f"Peak |dJ/dE| = {vmax:.4e}")

    # Wrap longitudes to [-180, 180] for plotting.
    lon_w = np.where(arr["lon"] > 180, arr["lon"] - 360, arr["lon"])
    lat = arr["lat"]

    # Log-scale colour, mask very weak cells.
    eps = vmax * 1e-4
    log_v = np.log10(np.maximum(abs_v, eps) / vmax)
    mask = log_v > -3
    print(f"Cells with |dJ/dE| > vmax/1e3: {int(mask.sum())}/{len(arr)}")

    # LA receptor lon/lat — pick the row matching (panel, i, j).
    la_row = arr[(arr["panel"] == la_panel) &
                 (arr["i"] == la_i) & (arr["j"] == la_j)]
    la_lon = la_row["lon"][0]
    la_lat = la_row["lat"][0]
    la_lon_w = la_lon - 360 if la_lon > 180 else la_lon

    # Try Cartopy for coastlines; fall back to plain matplotlib.
    try:
        import cartopy.crs as ccrs
        import cartopy.feature as cfeature
        fig = plt.figure(figsize=(14, 7))
        ax = plt.axes(projection=ccrs.PlateCarree())
        ax.set_global()
        ax.add_feature(cfeature.COASTLINE, linewidth=0.5, color="0.3")
        ax.add_feature(cfeature.BORDERS, linewidth=0.3, alpha=0.4, color="0.4")
        transform_kw = {"transform": ccrs.PlateCarree()}
        has_cartopy = True
    except ImportError:
        fig, ax = plt.subplots(figsize=(14, 7))
        ax.set_xlim(-180, 180)
        ax.set_ylim(-90, 90)
        ax.set_xlabel("Longitude [°E]")
        ax.set_ylabel("Latitude [°N]")
        ax.grid(True, alpha=0.3)
        transform_kw = {}
        has_cartopy = False

    # Plot all cells in faint background so the CS grid is visible.
    ax.scatter(lon_w, lat, s=4, c="0.85", alpha=0.5, edgecolors="none",
               **transform_kw)
    # Then the masked sensitivity cells on top.
    sc = ax.scatter(lon_w[mask], lat[mask], c=log_v[mask],
                    cmap="viridis", vmin=-3, vmax=0,
                    s=40, alpha=0.95, edgecolors="none",
                    **transform_kw)

    # LA receptor marker.
    ax.plot(la_lon_w, la_lat, marker="x", color="red",
            markersize=22, markeredgewidth=3, **transform_kw)
    ax.annotate(f"LA  ({la_lat:.1f}°N, {la_lon_w:.1f}°E)",
                (la_lon_w, la_lat),
                xytext=(la_lon_w - 50, la_lat + 12),
                color="red", fontsize=13, fontweight="bold",
                arrowprops=dict(arrowstyle="->", color="red", lw=1.2),
                **transform_kw)

    cbar = plt.colorbar(sc, ax=ax, orientation="horizontal",
                        pad=0.08, shrink=0.6)
    cbar.set_label("log₁₀(|dJ/dE| / max)  —  surface emission sensitivity")

    ax.set_title(
        f"6-hour LinRood adjoint footprint (backward)\n"
        f"Receptor: column-mean tracer at LA — C{Nc} CS grid, "
        f"{nsteps} substeps × 450 s = 6.0 h",
        fontsize=12,
    )

    out_path = CSV_PATH.parent / "linrood_la_footprint_c24_6h.png"
    fig.savefig(out_path, dpi=150, bbox_inches="tight")
    print(f"Saved plot to {out_path} ({out_path.stat().st_size / 1024:.1f} KB) [cartopy={has_cartopy}]")


if __name__ == "__main__":
    main()
